File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/python/dp_algorithms/rod_cutting.py
from __future__ import annotations

from functools import lru_cache

from dp_algorithms.common import DPResult


def memoized(prices: list[int], length: int) -> DPResult:
    @lru_cache(maxsize=None)
    def solve(remaining: int) -> int:
        if remaining == 0:
            return 0
        return max(prices[cut - 1] + solve(remaining - cut) for cut in range(1, remaining + 1))

    return DPResult("Rod Cutting", "memoization", solve(length))


def tabulated(prices: list[int], length: int) -> DPResult:
    dp = [0] * (length + 1)
    parent = [0] * (length + 1)
    for size in range(1, length + 1):
        for cut in range(1, size + 1):
            value = prices[cut - 1] + dp[size - cut]
            if value > dp[size]:
                dp[size] = value
                parent[size] = cut
    cuts: list[int] = []
    size = length
    while size > 0:
        cuts.append(parent[size])
        size -= parent[size]
    return DPResult("Rod Cutting", "tabulation", dp[length], {"cuts": cuts})


def space_optimized(prices: list[int], length: int) -> DPResult:
    result = tabulated(prices, length)
    return DPResult("Rod Cutting", "space optimized", result.value, result.details)

File: 29Items-Projects/7Algorithms/3-Dynamic-Programming/java/src/main/java/com/example/dp/RodCutting.java
package com.example.dp;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public final class RodCutting {
    private RodCutting() {
    }

    public static Result memoized(int[] prices, int length) {
        return new Result("Rod Cutting", "memoization", solve(prices, length, new HashMap<>()), Map.of());
    }

    private static int solve(int[] prices, int remaining, Map<Integer, Integer> memo) {
        if (remaining == 0) {
            return 0;
        }
        if (memo.containsKey(remaining)) {
            return memo.get(remaining);
        }
        int best = 0;
        for (int cut = 1; cut <= remaining; cut++) {
            best = Math.max(best, prices[cut - 1] + solve(prices, remaining - cut, memo));
        }
        memo.put(remaining, best);
        return best;
    }

    public static Result tabulated(int[] prices, int length) {
        int[] dp = new int[length + 1];
        int[] parent = new int[length + 1];
        for (int size = 1; size <= length; size++) {
            for (int cut = 1; cut <= size; cut++) {
                int value = prices[cut - 1] + dp[size - cut];
                if (value > dp[size]) {
                    dp[size] = value;
                    parent[size] = cut;
                }
            }
        }
        List<Integer> cuts = new ArrayList<>();
        for (int size = length; size > 0; size -= parent[size]) {
            cuts.add(parent[size]);
        }
        return new Result("Rod Cutting", "tabulation", dp[length], Map.of("cuts", cuts));
    }

    public static Result spaceOptimized(int[] prices, int length) {
        Result result = tabulated(prices, length);
        return new Result("Rod Cutting", "space optimized", result.value(), result.details());
    }
}
